livestream-param.ts

```diff
@@ -0,0 +1,70 @@
+import 'dotenv/config'
+import { RingApi } from '../api'
+import * as path from 'path'
+import { cleanOutputDirectory, outputDirectory } from './util'
+
+/**
+ * This example streams to files, each with 10 seconds of video.
+ * The output will be in output/part${part #}.mp4
+ **/
+
+async function example(myemail, mypassword,outputdir,duration) {
+  const ringApi = new RingApi({
+      // Replace with your ring email/password
+      email: myemail,
+      password: mypassword,
+      // Refresh token is used when 2fa is on
+      refreshToken: process.env.RING_REFRESH_TOKEN!,
+      debug: true
+    }),
+    [camera] = await ringApi.getCameras()
+
+  if (!camera) {
+    console.log('No cameras found')
+    return
+  }
+
+  //await cleanOutputDirectory()
+
+  console.log('Starting Video...')
+  const sipSession = await camera.streamVideo({
+    // save video 10 second parts so the mp4s are playable and not corrupted:
+    // https://superuser.com/questions/999400/how-to-use-ffmpeg-to-extract-live-stream-into-a-sequence-of-mp4
+    output: [
+      '-flags',
+      '+global_header',
+      '-f',
+      'segment',
+      '-segment_time',
+      '10', // 10 seconds
+      '-segment_format_options',
+      'movflags=+faststart',
+      '-reset_timestamps',
+      '1',

```

Comment:
One path you could pursue would be passing srtp options to ffmpeg here, which tells it where to pipe the audio/video to.  It's not an exact parallel, but you can see where I did something similar for _just audio_ in https://github.com/dgreif/ring/blob/25c77105fea32c68adfb9170c4b48a4debbe0dcb/homebridge/camera-source.ts#L240.  Assuming you can get Home Assistant to open ports for audio and video, you would just need to pass those port numbers in here and tell ffmpeg to pass the data along.

Reply:
that might just be the missing link as I was looking into what Home Assistant needs and it needs a stream! I'll investigate further :) Thanks again!

Reply:
@dgreif - just opened an 'issue' for you - hopefully you can expand your examples to include a stream example next to the 'stream-to-disk' example. thanks!

Reply:
Some time this seems to hang without return. I tried to hack around with this PR to my Home Assistant. My goal was just to save a video recording when there is a motion. I got it save and work when just trigger to save manually. However, when there is an actual motion detection, the node command seems to hang and never returned. I have to ``$ killall node`` to get out of it. Just a side note, my Ring username and password's naked in ``ps -ef | grep node`` :(

Reply:
hey @sjiampojamarn, thanks. I know this is a hacky solution - care to elaborate on how you got it to work in HA?

Reply:
@jeroenterheerdt, I followed most of your guide in README. Once we got `examples/livestream-param.js`, it's done, in general. We can run it from shell -- e.g. 
`./examples/livestream-param.js [username] [password] [outputdir] [duration]`

My HA's installed as a Python virtual env (https://www.home-assistant.io/docs/installation/virtualenv/). So in this case, I modified `homeassistant/lib/python3.6/site-packages/ring_doorbell/doorbot.py`
with the ``def get_livestream(self, js_file, outputdir, duration_seconds=30):`` function in this PR. 

In HA Ring component, `homeassistant/lib/python3.6/site-packages/homeassistant/components/ring`, I modified `camera.py` to accept and create entities when we have no Ring plan. In particular, this code block:

```
 if camera.has_subscription:
            cams.append(RingCam(hass, camera, config))
        else:
            cams.append(RingCam(hass, camera, config))
            #cams_no_plan.append(camera)
```

The snapshot and video of the Cam entity's taken from `_video_url` feeding to ffmpeg. So, I modified them -- e.g. 

```
        import glob
        ret = glob.glob("[outputdir]/*part0.mp4")
        ret.sort()
        self._video_url = ret[-1]
```
In the `__init__` of the RingCam class. And a bit in the `update`

```
#        if video_status == "ready" and (
#            self._last_video_id != last_recording_id or self._utcnow >= self._expires_at
#        ):

        #video_url = self._camera.recording_url(last_recording_id)
        import glob
        ret = glob.glob("[outputdir]/*part0.mp4")
        ret.sort()
        video_url = ret[-1]
```

Then, to expose HA service call I modified `homeassistant/lib/python3.6/site-packages/homeassistant/components/ring/__init__.py`

```
import asyncio
...
...
def service_camera_recording(service):
        camera_recording()

def camera_recording():
        """"Call service to record live camera to mp4."""
        _LOGGER.error("Recording Ring camera component")

        js_file = '[ring_js_script_dir]/livestream-param.js'
        outputdir = [outputdir]
        for doorbell in ring.doorbells:
            try:
                worker_task = asyncio.run_coroutine_threadsafe(
                    doorbell.get_livestream(js_file, outputdir, 25), hass.loop).result(40)
            except asyncio.TimeoutError:
                _LOGGER.error("Timeout recording " + doorbell)
                worker_task.cancel()
            except Exception as e:
                _LOGGER.error("Ring recording return: " + str(e))
            except:
                _LOGGER.error("Ring recoding return: ")

hass.services.register(DOMAIN, "recording", service_camera_recording)
```

My HA automation's simple. In `automations.yaml`, I added

```
- id: '123456789'
  alias: Ring Recording
  description: ''
  trigger:
  - entity_id: binary_sensor.ring_front_door_motion
    from: 'off'
    platform: state
    to: 'on'
  condition: []
  action:
  - service: ring.recording
```

Again, I think my solution here is even hackier than what you contributed here :). So far, I think it might be something to do with the node ring SIP component OR somehow with how Ring authentication and SIP session. I got into the problem of node not returning and hanging when:

  (1) time-to-time when I manually call Ring.recording service from HA. Mostly, the first time when I didn't interact with it for awhile (sleep -> active?)
  (2) If there is already on going Ring.recording running, any subsequent one returns very quickly without producing any output. The on-going one continue to finish the session and output correctly.
  (3) Never succeed so far when there is an actual Motion going on with Ring and ask it to call Ring.recording service.

Let me know. I'd really like to get this working : )

Reply:
One tiny thing, I added the new HA service in `homeassistant/lib/python3.6/site-packages/homeassistant/components/ring/services.yaml` as well:

```
update:
  description: Updates the data we have for all your ring devices
recording:
  description: Camera recording
```

Reply:
jeroenterheerdt@, I found a (hacky) solution to get motion, ding events locally saved and accessible thru HA. Somehow when Ring gets a motion or ding event, it rejects or at least makes the Sip session for recording gone crazy and never returned. At the end, I couldn't get it reliable/work to call it from HA with Python. So, I wrote/copied typescript and run separately with Node catching all events and ask it to record until it successful identified by Ring id. It seems the Ring id is the same for events occurred relatively at the same time period. Well, here is the ts file I made (or rather copies with modifies from the examples):

```
import { RingApi } from 'ring-client-api'
import * as path from 'path'

async function example(myemail,mypassword,outputdir,duration) {
  const ringApi = new RingApi({
      // Replace with your ring email/password
      email: myemail,
      password: mypassword,
      // Refresh token is used when 2fa is on
      // refreshToken: refresh_token,
      debug: true,
      cameraDingsPollingSeconds: 2
    }),
    [camera] = await ringApi.getCameras()

  if (!camera) {
    console.log('No cameras found')
    return
  }

  const express = require('express')
  const serveIndex = require('serve-index')
  const app = express()

  app.use(express.static(outputdir))
  app.use(serveIndex(outputdir, {'icons': true, 'view': 'details', 'fileList': 'name'}))
  app.use('/ring', express.static(outputdir))
  app.use('/ring', serveIndex(outputdir, {'icons': true, 'view': 'details', 'fileList': 'name'}))
  app.listen(3000, () => {
    console.log(
      'Listening on port 3000.'
    )
  })

  console.log('Doorbots connection: ' + camera.data.alerts.connection)
  var ring_event_id: string = ''
  camera.onNewDing.subscribe(ding => {
    const event =
      ding.kind == 'motion'
        ? 'motion'
        : ding.kind == 'ding'
        ? 'ding'
        : `Video started (${ding.kind})`
    console.log(
      `${event} on ${camera.name} camera. Ding id ${
        ding.id_str
      }.  Received at ${new Date()}`
    )
    const event_ts = new Date()
    const ring_event_id = ding.id_str
    const output_file = path.join(
      outputdir,
      event_ts.getFullYear() + "-" + event_ts.toLocaleDateString('en-US', {month: '2-digit'}) + "-" + event_ts.getDate() +
      "_" +
      event_ts.getHours().toString().padStart(2, '0') + "-" +
      event_ts.getMinutes().toString().padStart(2, '0') + "-" +
      ring_event_id +
      "_" + 'part0.mp4')
    const fs = require('fs')

    try {
      if (!fs.existsSync(output_file)) {
        console.log(output_file)
        camera.recordToFile(output_file, duration)
      } else {
        console.log('Already on Recording for: ' + output_file)
      }
    } catch(err) { 
      console.error(err)
    }
  })
  console.log('Listening for motion and doorbell presses on your cameras.')
}
// print process.argv
var args = process.argv.slice(2);
//email is args[0]
//password is args[1]
//outputdir is args[2]
//duration is args[3]
var email = args[0]
var password = args[1]
var outputdir = args[2]
var duration = args[3]

example(email, password, outputdir, duration)
```

So, motion and ding events are saved in [outputdir]. All the files are accessible (listed) thru http://localhost:3000. So, I put in HA Ring component with the above modified `camera.py` to show the latest video (and snapshot). Then use Panel iFrame in HA to access http://localhost:3000  for all the records. Well, actually, I use apache ProxyPass to map back the localhost:3000 to https web access point and put in HA (https) Panel iFrame. Looking forward a better, Python/HA native support.